query/src/main/java/org/infinispan/query/backend/TxQueryInterceptor.java

```diff
@@ -54,17 +58,18 @@ private InvocationStage commitModificationsToIndexFuture(InvocationContext ctx,
       final Map<Object, Object> oldValues = removed == null ? Collections.emptyMap() : removed;
 
       AbstractCacheTransaction transaction = txCtx.getCacheTransaction();
-      return asyncValue(CompletableFuture.allOf(transaction.getAllModifications().stream()
-            .filter(mod -> !mod.hasAnyFlag(FlagBitSets.SKIP_INDEXING))
-            .flatMap(mod -> mod.getAffectedKeys().stream())
-            .map(key -> {
+      CompletionStage<Void> stage = CompletionStages.performConcurrently(transaction.getAllModifications().stream()
+                  .filter(mod -> !mod.hasAnyFlag(FlagBitSets.SKIP_INDEXING))
+                  .flatMap(mod -> mod.getAffectedKeys().stream()).toList(), 100, Schedulers.from(new WithinThreadExecutor()),
```

Comment:
Instead of using toList, it might be worth it to add a performConcurrently method takes a Stream as the first parameter and invokes Flowable.fromStream and have a private method that encapsulates the parallel and other stuff for both.

Reply:
Absolutely! Fixing it